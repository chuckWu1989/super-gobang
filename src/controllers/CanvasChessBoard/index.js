const drawLine = (ctx, s, e) => {
  ctx.beginPath();
  ctx.moveTo(s[0], s[1]);
  ctx.lineTo(e[0], e[1]);
  ctx.closePath();
  ctx.stroke();
};
export const drawBoard = (canvas, props) => {
  const {
    grid,
    coord,
    width,
    ratio,
  } = props;
  const ctx = canvas.getContext('2d');
  const height = width * ratio;
  ctx.clearRect(0, 0, width, height);
  for (let i = 0; i <= (grid + 1); i += 1) {
    drawLine(ctx, coord[0][i], coord[grid + 1][i]);
    drawLine(ctx, coord[i][0], coord[i][grid + 1]);
  }
};

export default drawBoard;
