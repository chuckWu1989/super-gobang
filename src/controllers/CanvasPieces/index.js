import { radiusRatio, s, e } from './config';

const clearCanvas = (ctx, width, height) => {
  ctx.clearRect(0, 0, width, height);
};
const clearPiece = (ctx, coord, unitWidth) => {
  const cx = coord[0];
  const cy = coord[1];
  ctx.clearRect(cx - (unitWidth / 2), cy - (unitWidth / 2), unitWidth, unitWidth);
};
const drawPiece = (ctx, coord, unitWidth, color) => {
  const r = radiusRatio * (unitWidth / 2);
  const cx = coord[0];
  const cy = coord[1];
  ctx.beginPath();
  ctx.fillStyle = color;
  ctx.arc(cx, cy, r, s, e);
  ctx.fill();
  ctx.closePath();
};
const addOrRemovePiece = (ctx, nextProps, props) => {
  const { chessMoves: nextChessMoves, unitWidth, coord } = nextProps;
  const { chessMoves } = props;
  if (chessMoves.length < nextChessMoves.length) {
    const { x, y, color } = nextChessMoves[nextChessMoves.length - 1];
    const point = coord[x][y];
    drawPiece(ctx, point, unitWidth, color);
  } else if (chessMoves.length > nextChessMoves.length) {
    const { x, y } = chessMoves[chessMoves.length - 1];
    const point = coord[x][y];
    clearPiece(ctx, point, unitWidth);
  }
};
const redrawPieces = (ctx, props) => {
  const {
    width,
    height,
    unitWidth,
    coord,
  } = props;
  const { chessMoves } = props;
  clearCanvas(ctx, width, height);
  chessMoves.forEach((element) => {
    const { x, y, color } = element;
    const point = coord[x][y];
    drawPiece(ctx, point, unitWidth, color);
  });
};
export const detectChessMovesChange = (container, nextProps, props) => {
  const ctx = container.getContext('2d');
  const { chessMoves } = props;
  const { chessMoves: nextChessMoves } = nextProps;
  if (chessMoves.length !== nextChessMoves.length) {
    addOrRemovePiece(ctx, nextProps, props);
  }
};
export const detectWidthChange = (container, prevProps, props) => {
  const ctx = container.getContext('2d');
  const { width } = props;
  const { width: nextWidth } = prevProps;
  if (nextWidth !== width) {
    redrawPieces(ctx, prevProps);
  }
};

export default detectChessMovesChange;
