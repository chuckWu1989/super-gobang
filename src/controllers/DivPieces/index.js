import React from 'react';
import DivPiece from '../../views/DivPiece';
import { radiusRatio } from './config';

/**
 * Clear all content on chessboard
 * @param {array} children The children of component
 * @param {number} s Start point to be sliced
 * @param {number} e End point to be sliced
 * @returns {array} Updated children
 */
export function clearPiece(children, s, e) {
  const nextChildren = children.slice(s, e);
  return nextChildren;
}
/**
 * Draw piece on chessboard
 * @param {array} children The children of component
 * @param {array} coord The coordinate of piece
 * @param {number} unitWidth Unit width per grid
 * @param {string} color The color of piece
 * @returns {array} Updated children
 */
export const drawPiece = (children, coord, unitWidth, color) => {
  const size = radiusRatio * unitWidth;
  const x = (coord[0] - (size / 2)) + 1;
  const y = (coord[1] - (size / 2)) + 1;
  const key = `${x}-${y}-piece`;
  const piece = <DivPiece key={key} x={x} y={y} color={color} size={size} />;
  children.push(piece);
  return children;
};
/**
 * Rerender pieces on chessboard
 * @param {*} props The properties of the component
 */
export function redrawPieces(props) {
  const { unitWidth, coord, chessMoves } = props;
  const children = [];
  chessMoves.forEach((element) => {
    const { x, y, color } = element;
    const point = coord[x][y];
    drawPiece(children, point, unitWidth, color);
  });
  this.setState({ children });
}
/**
 * Add or remove piece on chessboard
 * @param {*} nextProps Next properties of the component
 * @param {*} props Current properties of the component
 */
export function addOrRemovePiece(nextProps, props) {
  const { chessMoves: nextChessMoves, unitWidth, coord } = nextProps;
  const { chessMoves } = props;
  const { children } = this.state;
  let nextChildren;
  if (chessMoves.length < nextChessMoves.length) {
    const { x, y, color } = nextChessMoves[nextChessMoves.length - 1];
    const point = coord[x][y];
    nextChildren = drawPiece(children, point, unitWidth, color);
  } else if (chessMoves.length > nextChessMoves.length) {
    const e = nextChessMoves.length - chessMoves.length;
    nextChildren = clearPiece(children, 0, e);
  }
  this.setState({ children: nextChildren.slice() });
}
/**
 * Detect if the chess moves
 * @param {object} nextProps Next properties of the component
 * @param {object} props Current properties of the component
 */
export function detectChessMovesChange(nextProps, props) {
  const { chessMoves: nextChessMoves } = nextProps;
  const { chessMoves } = props;
  if (nextChessMoves.length !== chessMoves.length) {
    addOrRemovePiece.call(this, nextProps, props);
  }
}
/**
 * Detect if the width of container is changed
 * @param {object} prevProps The previous properties of the component
 * @param {object} props Current properties of the component
 */
export function detectWidthChange(prevProps, props) {
  const { width } = props;
  const { width: prevWidth } = prevProps;
  if (prevWidth !== width) {
    redrawPieces.call(this, props);
  }
}

export default detectChessMovesChange;
