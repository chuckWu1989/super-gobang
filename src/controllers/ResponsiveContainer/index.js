import React from 'react';

export function setWidthAndHeight(container, nextProps) {
  return (
    function handler() {
      const props = nextProps !== undefined ? nextProps : this.props;
      const { children, ...others } = props;
      const childrenList = Array.isArray(children) ? children : [children];
      const width = container.offsetWidth;
      const height = container.offsetHeight;
      const clonedChildren = childrenList.map((child, idx) => {
        const key = idx;
        return React.cloneElement(child, {
          ...others,
          width,
          height,
          key,
        });
      });
      this.setState({ children: clonedChildren });
    }
  );
}
export function addScreenListener(container) {
  let handler = setWidthAndHeight(container);
  handler = handler.bind(this);
  window.addEventListener('resize', handler);
  handler();
  return handler;
}
export const removeScreenListener = (handler) => {
  window.removeEventListener('resize', handler);
};

export default addScreenListener;
