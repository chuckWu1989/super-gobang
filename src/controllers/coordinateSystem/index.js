export function setCoordinate() {
  const { width, ratio, grid } = this.props;
  if (typeof width !== 'number') {
    return false;
  }
  const coord = new Array(grid + 2);
  const trueWidth = Math.floor(width / 15) * 15;
  const trueHeight = Math.floor((trueWidth * ratio) / 15) * 15;
  const unitWidth = trueWidth / (grid + 1);
  const unitHeight = trueHeight / (grid + 1);
  for (let x = 0; x <= grid + 1; x += 1) {
    coord[x] = new Array(grid);
    for (let y = 0; y <= grid + 1; y += 1) {
      coord[x][y] = [unitWidth * x, unitHeight * y];
    }
  }
  return this.setState({ coord, width: trueWidth, height: trueHeight });
}

export default setCoordinate;
