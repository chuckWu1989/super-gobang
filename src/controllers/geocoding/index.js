/**
 * Calculate coordinates and set into state.
 */
export function setCoordinate() {
  const { width, ratio, grid } = this.props;
  if (typeof width !== 'number') {
    return false;
  }
  const coord = new Array(grid + 2);
  const unitWidth = Math.round(width / (grid + 1));
  const unitHeight = Math.round((width * ratio) / (grid + 1));
  const trueWidth = unitWidth * (grid + 1);
  const trueHeight = unitHeight * (grid + 1);
  for (let x = 0; x <= grid + 1; x += 1) {
    coord[x] = new Array(grid);
    for (let y = 0; y <= grid + 1; y += 1) {
      coord[x][y] = [unitWidth * x, unitHeight * y];
    }
  }
  return this.setState({
    coord,
    unitWidth,
    unitHeight,
    width: trueWidth,
    height: trueHeight,
  });
}
/**
 * Detect if the width of container is changed
 * @param {*} nextProps Next properties of the component
 * @param {*} props Current properties of the component
 */
export function detectWidthChanged(nextProps, props) {
  const { width: nextWidth } = nextProps;
  const { width } = props;
  if (nextWidth !== width) {
    setCoordinate.call(this);
  }
}

export default setCoordinate;
