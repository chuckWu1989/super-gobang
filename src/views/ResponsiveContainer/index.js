/* eslint react/no-did-mount-set-state: off */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './style.less';

class ResponsiveContainer extends Component {
  static defaultProps = {
    updateChildren: () => {},
    addScreenListener: () => {},
    removeScreenListener: () => {},
  }
  static propTypes = {
    updateChildren: PropTypes.func,
    addScreenListener: PropTypes.func,
    removeScreenListener: PropTypes.func,
  }
  constructor(props) {
    super(props);
    this.state = { hanlder: undefined, children: null };
  }
  componentDidMount() {
    const hanlder = this.props.addScreenListener.call(this, this.container);
    this.setState({ hanlder });
  }
  componentWillReceiveProps(nextProps) {
    this.props.updateChildren.call(this, nextProps);
  }
  componentWillUnmount() {
    this.props.removeScreenListener(this.state.hanlder);
  }
  render() {
    const { children } = this.state;
    return (
      <div
        className="responsive-container-style"
        ref={(self) => { this.container = self; }}
      >
        {children}
      </div>
    );
  }
}

export default ResponsiveContainer;
