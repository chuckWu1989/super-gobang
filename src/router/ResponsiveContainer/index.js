import { addScreenListener, removeScreenListener, updateChildren } from '../../controllers/ResponsiveContainer';
import ResponsiveContainer from '../../views/ResponsiveContainer';
import { connect } from '../../hocs/connect';

export const method = () => ({
  addScreenListener,
  removeScreenListener: handler => removeScreenListener(handler),
  updateChildren,
});

export default connect(method)(ResponsiveContainer);
