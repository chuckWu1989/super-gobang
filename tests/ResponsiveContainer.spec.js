import React from 'react';
import toJson from 'enzyme-to-json';
import ResponsiveContainer from '../src/views/ResponsiveContainer';
import { method } from '../src/router/ResponsiveContainer';
import * as controller from '../src/controllers/ResponsiveContainer';

const { shallow } = global;
const props = {};

describe('ResponsiveContainer', () => {
  describe('view', () => {
    it('should be rendered properly', () => {
      const tree = toJson(shallow(<ResponsiveContainer {...props} />));
      expect(tree).toMatchSnapshot();
    });
    it('should call methods during lifecycle', () => {
      const ownProps = {
        ...props,
        updateChildren: jest.fn(),
        addScreenListener: jest.fn(),
        removeScreenListener: jest.fn(),
      };
      const wrapper = shallow(
        <ResponsiveContainer {...ownProps} />,
        { lifecycleExperimental: true },
      );
      wrapper.setProps({ rand: '123' });
      wrapper.unmount();
      expect(ownProps.updateChildren).toHaveBeenCalledTimes(1);
      expect(ownProps.addScreenListener).toHaveBeenCalledTimes(1);
      expect(ownProps.removeScreenListener).toHaveBeenCalledTimes(1);
    });
  });
  describe('router', () => {
    it('should pass addScreenListener', () => {
      expect('addScreenListener' in method()).toBeTruthy();
    });
    it('should pass removeScreenListener', () => {
      expect('removeScreenListener' in method()).toBeTruthy();
    });
    it('should pass updateChildren', () => {
      expect('updateChildren' in method()).toBeTruthy();
    });
  });
  describe('controller', () => {
    describe('setWidthAndHeight', () => {
      const container = { offsetWidth: 150, offsetHeight: 150 };
      it('should return handler function', () => {
        const result = controller.setWidthAndHeight();
        expect(typeof result).toBe('function');
      });
      it('should update children if chilren is defined and children is not an array', () => {
        let result;
        const ownProps = {
          props: { children: <div /> },
          setState: (state) => { result = state; },
        };
        const handler = controller.setWidthAndHeight(container);
        handler.call(ownProps);
        expect(result).toBeDefined();
      });
      it('should update children if children is defined and children is an array', () => {
        let result;
        const ownProps = {
          props: { children: [<div />] },
          setState: (state) => { result = state; },
        };
        const handler = controller.setWidthAndHeight(container);
        handler.call(ownProps);
        expect(result).toBeDefined();
      });
      it('should skip updating if children is undefined', () => {
        let result;
        const ownProps = { props, setState: (state) => { result = state; } };
        const handler = controller.setWidthAndHeight(container);
        handler.call(ownProps);
        expect(result).toBeUndefined();
      });
    });
    describe('updateChildren', () => {
      let result;
      let ownProps;
      beforeEach(() => {
        result = undefined;
        ownProps = {
          props,
          state: { children: [<div key="test" width={100} height={100} />] },
          setState: (state) => { result = state; },
        };
      });
      it('should update children if children is not an array', () => {
        const nextProps = {
          children: <div />,
        };
        controller.updateChildren.call(ownProps, nextProps);
        expect(result).toBeDefined();
      });
      it('should update children if children is an array', () => {
        const nextProps = {
          children: [<div />],
        };
        controller.updateChildren.call(ownProps, nextProps);
        expect(result).toBeDefined();
      });
    });
    describe('addScreenListener', () => {
      it('should return hanlder', () => {
        let result;
        const container = { offsetWidth: 150, offsetHeight: 150 };
        const ownProps = {
          props: { children: <div /> },
          setState: (state) => { result = state; },
        };
        const temp = window.addEventListener;
        window.addEventListener = jest.fn();
        controller.addScreenListener.call(ownProps, container);
        expect(result).toBeDefined();
        expect(window.addEventListener).toHaveBeenCalledTimes(1);
        window.addEventListener = temp;
      });
    });
    describe('removeScreenListener', () => {
      it('should call removeEventListener method', () => {
        const temp = window.removeEventListener;
        window.removeEventListener = jest.fn();
        controller.removeScreenListener();
        expect(window.removeEventListener).toHaveBeenCalledTimes(1);
        window.removeEventListener = temp;
      });
    });
  });
});
