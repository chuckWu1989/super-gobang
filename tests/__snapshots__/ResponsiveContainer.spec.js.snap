// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponsiveContainer view should be rendered properly 1`] = `
<div
  className="responsive-container-style"
/>
`;
